Feat/split terms

eddiesshop opened the pull request:
This command will help identify duplicate slugs on any WordPress site. It then provides an interactive experience for handling those duplicate slugs, allowing for the user to selectively choose which slugs should be handled by the command.

The way the command "handles" the duplicate slugs is in 2 ways:

1. There are the same number of Term and Taxonomy IDs, so the first Term ID remains as the original slug, and subsequent Term slugs are suffixed with -1, -2, -3, etc.
2. There are more than one Taxonomy IDs for the same Term ID. In this case, new Term IDs are created with uniquely numbered slugs (suffixed -1, -2, -3, etc) and then the Taxonomy ID is updated with the newly created Term ID.

iuravic commented:
> subsequent Term slugs are suffixed with -1, -2, -3, etc
and
> new Term IDs are created with uniquely numbered slugs (suffixed -1, -2, -3, etc)

@eddiesshop , I wonder is there a need to change slugs for newly created terms?

I just did a test on a fresh blank WP page, created a new tag called "blue", and then a new category called "blue",
![image](https://user-images.githubusercontent.com/29167323/224306914-6a42c18c-28ae-41b0-a750-22efa8ac8ca7.png)
![image](https://user-images.githubusercontent.com/29167323/224306943-1b2fea32-41c3-4172-8a14-f252e49b41ca.png)

Both these "blue"s have same slugs and seem to work fine. So should we be updating slugs when splitting them?

iuravic commented:
@eddiesshop I created a blank test WP site with these terms and taxonomies

`wp_terms table`

| term_id | name               | slug               |
| ------- |:------------------:|:------------------:|
| 2       | blue               | blue               |
| 3       | dog                | dog                |
| 4       | vegetable          | vegetable          |
| 5       | notsharedcat       | notsharedcat       |
| 6       | blue               | blue               |
| 7       | dog                | dog                |
| 8       | vegetable          | vegetable          |
| 9       | notsharedcustomtax | notsharedcustomtax |

`wp_term_taxonomy table`

| term_taxonomy_id | term_id | taxonomy      |
| ---------------- |:-------:|:-------------:|
| 1                | 1       | category      |
| 2                | 2       | category      |
| 3                | 3       | category      |
| 4                | 4       | category      |
| 5                | 5       | category      |
| 6                | 6       | post_tag      |
| 7                | 7       | post_tag      |
| 8                | 8       | post_tag      |
| 9                | 9       | testcustomtax |
| 10               | 8       | testcustomtax |
| 11               | 7       | testcustomtax |

And here's what the command is reporting:
![image](https://user-images.githubusercontent.com/29167323/224320403-0dbd5fa5-e280-42c1-9a0c-06ad862457fd.png)

Could you please check if this is correct?

Seems to me that it should  be reporting these duplicates:
- slug `dog`, taxonomies `post_tag, testcustomtax`, term_id_count `2`, term_taxonomy_id_count `2`
- slug `vegetable`, taxonomies `post_tag, testcustomtax`, term_id_count `2`, term_taxonomy_id_count `2`

It shouldn't really be reporting `dog` because those are two separate term_ids `3, 7` only used once in wp_term_taxonomy with term_taxonomy_ids `3, 11`.

Other than this, I think it would be super useful if the table above would list actual term_ids and term_taxonomy_ids for terms, because it would make debugging much easier -- I could see from the table which specific terms and term_taxonomy_ids are going to be affected.

iuravic commented:
@eddiesshop here's my test site's [full DB dump](https://drive.google.com/file/d/1ONCrOaOG3Qne1BRpWwVqi6RWLY-fHaSY/view?usp=share_link) with these terms, taxonomies and 5 different posts I assigned these to differently.

iuravic reviewed (changes requested):
Added comments [here](https://github.com/Automattic/newspack-custom-content-migrator/pull/267#issuecomment-1463684469), [here](https://github.com/Automattic/newspack-custom-content-migrator/pull/267#issuecomment-1463765198), and [here](https://github.com/Automattic/newspack-custom-content-migrator/pull/267#issuecomment-1463768185).

eddiesshop commented:
@iuravic thanks for the detailed analysis you performed with this branch!

I think your comments make total sense, especially given the default behavior WordPress exhibits when creating these taxonomies.

I modified the query so that all taxonomies are now shown, including custom ones (previously it was limited to just categories and post tags). I also added another virtual column which displays the term_id -> term_taxonomy_id -> taxonomy relationship. I did it in the query so as to not have to do any PHP processing on the result set.

Let me know what you think about these changes!

iuravic reviewed (changes requested):
The command is throwing a fatal DB error and it doesn't work for me. I'm using the same DB dump I attached -- please try it again. Here's trying to run it with several modes -- first just display, then run, then interactive, then finally display again, and no changes get done. Plus there's a DB error displayed.

```
bash-3.2$ wp newspack-content-migrator split-duplicate-term-slugs --display
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
| slug      | taxonomies               | term_id:term_taxonomy_id | term_id_count | term_taxonomy_id_count        |
|           |                          | :taxonomy                |               |                               |
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
| vegetable | post_tag, testcustomtax  | 8:8:post_tag  |  8:10:te | 1             | 2                             |
|           |                          | stcustomtax              |               |                               |
| dog       | post_tag, testcustomtax  | 7:7:post_tag  |  7:11:te | 1             | 2                             |
|           |                          | stcustomtax              |               |                               |
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
bash-3.2$ wp newspack-content-migrator split-duplicate-term-slugs
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
| slug      | taxonomies               | term_id:term_taxonomy_id | term_id_count | term_taxonomy_id_count        |
|           |                          | :taxonomy                |               |                               |
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
| vegetable | post_tag, testcustomtax  | 8:8:post_tag  |  8:10:te | 1             | 2                             |
|           |                          | stcustomtax              |               |                               |
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
Splitting duplicate term slug...
bash-3.2$ wp newspack-content-migrator split-duplicate-term-slugs --interactive
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
| slug      | taxonomies               | term_id:term_taxonomy_id | term_id_count | term_taxonomy_id_count        |
|           |                          | :taxonomy                |               |                               |
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
| vegetable | post_tag, testcustomtax  | 8:8:post_tag  |  8:10:te | 1             | 2                             |
|           |                          | stcustomtax              |               |                               |
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
<div id="error"><p class="wpdberror"><strong>WordPress database error:</strong> [Table &#039;local.wp_terms&#039; doesn&#039;t exist]<br /><code>SELECT
	                t.term_id,
	                t.name, t.slug,
	                tt.term_taxonomy_id,
	                tt.taxonomy,
	                tt.parent,
	                tt.count
				FROM wp_terms t
				INNER JOIN wp_term_taxonomy tt ON t.term_id = tt.term_id
				WHERE t.slug = &#039;vegetable&#039;
				AND tt.taxonomy IN ( &quot;category&quot;, &quot;post_tag&quot; )
				ORDER BY t.term_id</code></p></div>+
|
+
+
What would you like to do with this slug? s(p)lit term and rename slug, (s)kip, return to (a)uto, or (q)uit?: a
Splitting duplicate term slug...
bash-3.2$ wp newspack-content-migrator split-duplicate-term-slugs
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
| slug      | taxonomies               | term_id:term_taxonomy_id | term_id_count | term_taxonomy_id_count        |
|           |                          | :taxonomy                |               |                               |
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
| vegetable | post_tag, testcustomtax  | 8:8:post_tag  |  8:10:te | 1             | 2                             |
|           |                          | stcustomtax              |               |                               |
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
Splitting duplicate term slug...
bash-3.2$ wp newspack-content-migrator split-duplicate-term-slugs --display
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
| slug      | taxonomies               | term_id:term_taxonomy_id | term_id_count | term_taxonomy_id_count        |
|           |                          | :taxonomy                |               |                               |
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
| vegetable | post_tag, testcustomtax  | 8:8:post_tag  |  8:10:te | 1             | 2                             |
|           |                          | stcustomtax              |               |                               |
| dog       | post_tag, testcustomtax  | 7:7:post_tag  |  7:11:te | 1             | 2                             |
|           |                          | stcustomtax              |               |                               |
+-----------+--------------------------+--------------------------+---------------+-------------------------------+
bash-3.2$
```

eddiesshop commented:
I apologize for the DB errors! I didn't test this branch with custom taxonomy, and so that was causing errors which weren't expected and weren't handled properly. I believe the branch is now ready for further testing!

iuravic reviewed (changes requested):
I'm finding one issue,
- termmeta seems to be completely lost for newly split terms.

Still wondering about this question asked above,
- [question](https://github.com/Automattic/newspack-custom-content-migrator/pull/267#issuecomment-1463684469) regarding slugs

And two small suggestions,
- I'd recommend that we now removed the second column because the third column now contains that same info,
![image](https://user-images.githubusercontent.com/29167323/225003743-26c1d942-2c43-4a41-90ed-dd974c9c59fb.png)
- I would also recommend that we changed the wording from "Splitting duplicate term slug..." to "Splitting duplicate terms..." since the major change done is that terms are being split, while whether their slugs are preserved or changed seems secondary to me

eddiesshop commented:
Hey Ivan! Thanks for taking another look at the PR. Please see below for my responses to your concerns.

>I'm finding one issue,
>
>termmeta seems to be lost for newly split terms

Yes that would make sense since we are creating a brand new term with a unique slug. How would you like me to handle this? Should I just duplicate whatever `wp_termmeta` rows exist for the original term/slug to the new term/slug?

>Still wondering about this question asked above,
>
> - [question](https://github.com/Automattic/newspack-custom-content-migrator/pull/267#issuecomment-1463684469) regarding slugs

Even though default WordPress behavior is to allow duplicate term-slugs for *categories and post tags*, it is still essential to de-dupe the slugs when they are linked across a category or post tag. Please see the following example for the reason why:

### All taxonomies linked to `slug = dog`
```mysql
SELECT 
    t.term_id, 
    t.name, 
    t.slug, 
    tt.term_taxonomy_id, 
    tt.taxonomy 
FROM wp_zyv6o2ibm4_terms t 
    INNER JOIN wp_zyv6o2ibm4_term_taxonomy tt ON t.term_id = tt.term_id 
WHERE t.slug = 'dog';
```

![image](https://user-images.githubusercontent.com/17132385/225137041-745b33bc-35bf-4f67-8eab-b596fa63a591.png)

### Selecting `dog` as a tag for post
![image](https://user-images.githubusercontent.com/17132385/225138111-e01efe0f-d09e-4ae0-a704-4edeb7331381.png)

### WordPress selects the wrong taxonomical record when assigning post tag
```mysql
SELECT 
       p.ID, 
       p.post_name, 
       tr.term_taxonomy_id, 
       tt.taxonomy, 
       t.term_id, 
       t.name, 
       t.slug 
FROM wp_zyv6o2ibm4_term_relationships tr
    LEFT JOIN wp_zyv6o2ibm4_term_taxonomy tt ON tr.term_taxonomy_id = tt.term_taxonomy_id
    INNER JOIN wp_zyv6o2ibm4_terms t ON t.term_id = tt.term_id
    INNER JOIN wp_zyv6o2ibm4_posts p ON tr.object_id = p.ID
WHERE p.ID = 20
ORDER BY p.ID, t.slug;
```
![image](https://user-images.githubusercontent.com/17132385/225138436-b2086269-9b98-49dc-ae0f-46051da7e7a8.png)

Because of this behavior, it is necessary to ensure that a 1-to-1 relationship is maintained between Term -> Slug -> and Taxonomy. This is where this PR comes into play.

### After splitting the term-slug
![image](https://user-images.githubusercontent.com/17132385/225141983-4b25b92d-46f4-4c7e-8217-6c22c834c94d.png)

![image](https://user-images.githubusercontent.com/17132385/225142161-bb7dbdf1-794b-4d57-82da-515dff503c1c.png)

```mysql
SELECT 
       t.term_id, 
       t.name, 
       t.slug, 
       tt.term_taxonomy_id, 
       tt.taxonomy
FROM wp_zyv6o2ibm4_terms t
    INNER JOIN wp_zyv6o2ibm4_term_taxonomy tt ON t.term_id = tt.term_id
WHERE t.slug LIKE 'dog%';
```
![image](https://user-images.githubusercontent.com/17132385/225143277-aa87666a-2fe2-4276-8bd2-469fd9ceaac6.png)

```mysql
SELECT
       p.ID,
       p.post_name,
       tr.term_taxonomy_id,
       tt.taxonomy,
       t.term_id,
       t.name,
       t.slug
FROM wp_zyv6o2ibm4_term_relationships tr
    LEFT JOIN wp_zyv6o2ibm4_term_taxonomy tt ON tr.term_taxonomy_id = tt.term_taxonomy_id
    INNER JOIN wp_zyv6o2ibm4_terms t ON t.term_id = tt.term_id
    INNER JOIN wp_zyv6o2ibm4_posts p ON tr.object_id = p.ID
WHERE p.ID = 23
ORDER BY p.ID, t.slug;
```
![image](https://user-images.githubusercontent.com/17132385/225143495-d42ba114-c18f-40c7-8f0a-9b2caf68a3ca.png)

>Regarding term_id_count,
>
>I'm not sure I understand term_id_counts meaning -- can you please share in words what term_id_count means?
it seems to be always showing value 1. Is that supposed to be so?

We can remove this column now actually. Because we are now grouping slugs across term IDs and not taxonomies, the value for that column will always be 1.

>I'd recommend that we now removed the second column because the third column now contains that same info,

I disagree with this because this column would provide a quick glance at the taxonomies involved with the operation. However, I will make a change that omits that column by default, but can be added with a flag.

>I would also recommend that we changed the wording from "Splitting duplicate term slug..." to "Splitting term..." since the major change done is that terms are being split, while whether their slugs are preserved or changed seems secondary to me

Sure thing 👍

iuravic commented:
> Yes that would make sense since we are creating a brand new term with a unique slug. How would you like me to handle this? Should I just duplicate whatever `wp_termmeta` rows exist for the original term/slug to the new term/slug?

Yeah, I think so. Just duplicating all the termmetas for newly created terms should do the trick 👍 

> Even though default WordPress behavior is to allow duplicate term-slugs for categories and post tags, it is still essential to de-dupe the slugs when they are linked across a category or post tag. Please see the following example for the reason why ...

Thanks so much Super clear to me now! 🙇

> > I'd recommend that we now removed the second column because the third column now contains that same info,
> I disagree with this because this column would provide a quick glance at the taxonomies involved with the operation. However, I will make a change that omits that column by default, but can be added with a flag.

That's fine, if you find it useful, let's leave it.

iuravic reviewed (changes requested):
Please see my latest reply [here](https://github.com/Automattic/newspack-custom-content-migrator/pull/267#issuecomment-1469858984). We just need termmeta fixed and I think that will be all ;)

So happy to have this command finally, it will be very useful.

eddiesshop commented:
Hey @iuravic ! I committed a change to duplicate termmeta when a term is split (i.e. when a new term record is created). Please review the change when you get a chance!

iuravic reviewed (changes requested):
This works well but doesn't fully migrate all termmeta. WP supports multiple meta with same meta_key, all of the values/rows/array elements should be synced, not just `$meta_value[0]`.

iuravic commented:
@eddiesshop , please run PHPCS as well since this is a reusable migrator, it's presently throwing errors. Make sure those are cleared, especially SQL param sanitization using `$wpdb->prepare()` for all statements.

You'll also find errors from previous commits. We should aim to clean the code we've previously worked on, plus if it's easy and quick to fix remaining errors, that would be appreciated too.
